### PR TITLE
Allowing Hierarchical Non Centered Parametrization

### PR DESCRIPTION
**Pull request by cetagostini**
## Related Issue

- [x] Closes #743 
- [ ] Related to #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [x] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [x] MMM
- [ ] CLV

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--747.org.readthedocs.build/en/747/

### CODE REVIEW

File: pymc_marketing/model_config.py [hunk 1]
```diff
@@ -388,13 +466,12 @@ def create_distribution(
     **kwargs
         Additional keyword arguments for the distribution.
 
-    Returns
-    -------
-    TensorVariable
-        A PyMC random variable.
-
```
Comment (juanitorduz) [inline pymc_marketing/model_config.py:395]: why did you erase this?

**Review by juanitorduz — COMMENTED** [review]
Just some little questions

File: pymc_marketing/model_config.py [hunk 2]
```diff
@@ -369,6 +388,65 @@ def handle_parameter_distributions(
...
+def create_hierarchical_non_center(
+    name: str,
+    distribution_kwargs: dict[str, Any],
+    **kwargs,
+) -> pt.TensorVariable:
+    """
+    Create a hierarchical non-centered distribution.
+
+    This function constructs a hierarchical non-centered distribution using the provided
+    distribution parameters for offset, mu, and sigma. It returns a deterministic variable
```
Comment (juanitorduz) [inline pymc_marketing/model_config.py:400]: Why do we need this offset random variable? Usually the non-center parametrization in the Normal distribution case means using a standard normal  multiply it with the variance and add the mean as in https://www.pymc.io/projects/examples/en/latest/generalized_linear_models/multilevel_modeling.html#non-centered-parameterization Can you provide some explanation in the doctrines?

**Comment by juanitorduz** [comment]
@cetagostini, I do not see any real change in the mmm notebook, right (or did I miss something)? If so, can you remove it from this PR?

**Comment by cetagostini** [comment]
> @cetagostini, I do not see any real change in the mmm notebook, right (or did I miss something)? If so, can you remove it from this PR?

Yes, previous notebook has "likelihood sigma", now we replace to be call "y". 

I update the notebook, but we should update all. I was playing with TVP, and same thing!

File: pymc_marketing/model_config.py (hunk 2 above)
Comment (wd60622) [reply pymc_marketing/model_config.py:400]: +1 and I have a counter example for why this is wrong. It's pretty clear with 2D case

File: pymc_marketing/model_config.py (hunk 1 above)
Comment (cetagostini) [reply pymc_marketing/model_config.py:395]: Mistake, adding this back!

File: pymc_marketing/model_config.py (hunk 2 above)
Comment (cetagostini) [reply pymc_marketing/model_config.py:400]: @juanitorduz How I see it, when we use Normal(0,1) then the global mean is the baseline (mu), making the model simpler to interpret and ensuring that the deviations are centered around the global mean. So, global_mu + offset + global_sigma will be the posterior parameter for the each specific hierarchy.
```
#non-centered
offset = pm.Normal(
  'offset', 
  mu=0, sigma=1
  dims=("geo")
 )
  mu_global = pm.HalfNormal(name="mu_global",sigma=2)
  sigma_global = pm.HalfNormal(name="sigma_global",sigma=1)

  vmax = pm.Deterministic(
    "vmax", 
    var=mu_global + offset * sigma_global, 
    dims=("geo")
)
```

This as well mean, offset controls how much each geo deviates from the global mean. @twiecki mentioned in their article:

> Instead of saying that our individual slopes b are normally distributed around a group mean (i.e. modeling their absolute values directly), we can say that they are offset from a group mean by a certain value (b_offset; i.e. modeling their values relative to that mean).

So, by having the prior distribution where the offset equal 0, we basically say that it does not deviate from the global mean mean. Example picture:
image

In the opposite case, offset becomes how much we deviate. And depending on how much we deviate, the final parameter will have a value, Example, the following image where offset is different from zero, but is still a normal distribution:
image

If our offset is set for a non-normal distribution with a non-zero mean (e.g., gamma distribution) shifts the global mean of the parameters but does not inherently lead to a wrong model. It changes the interpretation.

Example: Using `Gamma` because it's cap to zero, transform the global mean into a global minimum. So, our posterior distribution for the parameter could not have a value lower than that.

image

This is not directly a wrong model, it's just a model with a different interpretation (Which I accept could be not common). But it's a choice and depending the parameter to inform, and your problem, you could use it or not.

Happy to be challenged :) 

cc: @wd60622

Comment (wd60622) [reply pymc_marketing/model_config.py:400]: My biggest issue is that this still requirements on the user to know how to correctly provide the dims of offset.

## 1D Case
Even in the 1D, the user can deviate from the actual

```python
distribution = {
    "dist": "Normal", 
    "kwargs": {
        "mu": {"dist": "Normal", "kwargs": {"mu": 0, "sigma": 1}}, 
        "sigma": {"dist": "HalfNormal", "kwargs": {"sigma": 1}}, 
    }, 
    "dims": ("channel"), 
}
```

Since there are "channel" draws from the parent normal distributions, this in my mind is only one option for non-centered. That is: 

```python
distribution = {
    "dist": "Normal", 
    "kwargs": {
        "offset": {
            "dist": "Normal", 
            "kwargs": {"mu": 0, "sigma": 1}, 
            # This has to be center dims in order to get the correct number of draws
            "dims": "channel",
        }, 
        "mu": {"dist": "Normal", "kwargs": {"mu": 0, "sigma": 1}}, 
        "sigma": {"dist": "HalfNormal", "kwargs": {"sigma": 1}}, 
    }, 
    "dims": ("channel"), 
}
```

But user could have offset.dims = () or use mu.dims = channel instead? There is a lot of room for error

## 2D Case

This centered parameterization assumes "channel" x "geo" draws from the Normal

```python
distribution = {
    "dist": "Normal", 
    "kwargs": {
        "mu": {"dist": "Normal", "kwargs": {"mu": 0, "sigma": 1}, "dims": channel"}, 
        "sigma": {"dist": "HalfNormal", "kwargs": {"sigma": 1}, "dims": "channel"}, 
    }, 
    "dims": ("channel", "geo"),
}
```

However, I need to be explicit about offset if I want that! But that is not required to be so which puts the burden onto the user 😢 

For instance, the user could do: 

- offset.dims = "geo"
- offset.dims = "channel"
- offset.dims = ()
- offset.dims = ("channel", "geo") # Correct number of draws in my mind

The offset.dims needs to be the correct number of draws or you end up just broadcasting vectors which doesn't give the write answer

**Review by wd60622 — CHANGES_REQUESTED** [review]
Left comment about issue with dims in the thread

File: pymc_marketing/model_config.py (hunk 2 above)
Comment (cetagostini) [reply pymc_marketing/model_config.py:400]: Sure, this make sense but we are talking about an user usability issue. 

I feel not allowed is to radical, we are basically saying, "better not to give driver license to anyone because a few drivers could be involved in accidents".  

Nevertheless, I have a few things which could work to solve this, based on what you said, we just need to highlight a warning or error when final parameter and offset has different dimensions.

We can explain as well in the notebook, I really don't see this as a dramatic issue!

Comment (juanitorduz) [reply pymc_marketing/model_config.py:400]: Great discussion 💪 ! What if we, for this first iteration, simply use the flag `"centered": False / True` and get feedback and iterate ?

Comment (cetagostini) [reply pymc_marketing/model_config.py:401]: I just did an small changes to raise warning and overwrite the dims in case don't match. This should solve the issue, as well should we get better feedback no?

If we just move to `"centered": False / True` will always work and user will never play with anything else, so I don't feel any feedback will come. Adding the warning, they could share if they are playing with the distributions, and getting warnings or not. Which, I believe is more constructive. no?

cc: @juanitorduz @wd60622

Comment (cetagostini) [reply pymc_marketing/model_config.py:401]: Final point about not allowing moving the offset. If the user wants to make `beta` a non-centered hierarchical parameter, and we have offset always `Normal`. It could have negative values ​​even if global `mu` and `sigma` are positive only.

I think that would force us to then have to make an adjustment when they want to force positive values. But if we can play with all the distributions they can make all the distributions Gamma, HalfNormal, TruncatedNormal and that's it. :)

Comment (juanitorduz) [reply pymc_marketing/model_config.py:401]: >  if we just move to "centered": False / True will always work and user will never play with anything else,

For this PR, and given the user base we have now, I think is a wise bet. We can do more experiments and test them in a different PR. I would like, for example, to test edge cases. But having this boolean flag can help our users with divergences which can be avoidable by this easy parametrization change. WDYT?

Comment (cetagostini) [reply pymc_marketing/model_config.py:401]: > I would like, for example, to test edge cases.

Sure, are you talking about model flexibility? or mathematical properties? Based on those cases, we can usually cover those on the unit tests.

> But having this boolean flag can help our users with divergences which can be avoidable by this easy parametrization change.

Totally agree, but then are we discussing a line of code? 😅 because the same can be achieve if:

```python
"kwargs": {
        "mu": {"dist": "Normal", "kwargs": {"mu": 0, "sigma": 1}}, 
        "sigma": {"dist": "HalfNormal", "kwargs": {"sigma": 1}}, 
    }
   . . .
"centered": False,
```
or

```python
"kwargs": {
        "offset": {"dist": "Normal", "kwargs": {"mu": 0, "sigma": 1}},
        "mu": {"dist": "Normal", "kwargs": {"mu": 0, "sigma": 1}}, 
        "sigma": {"dist": "HalfNormal", "kwargs": {"sigma": 1}}, 
    }
   . . .
```

Both cases is a single extra line of code, and using the `default` users get the same output.

For me the conversation should be around:
a) What is more flexible?
b) What is more sustainable?
c) What aligns with our final goal?

As I see it, the idea is our users have an API that is flexible enough to build different models, where **imagination is the limit**. Limiting it to a single type of parameterization would lead us to find particular cases that are not addressed and create solutions to handle those (Example, non-negative parameters from the top of my head, more will appear for sure).

On the other hand we can leave the configuration flexible, by `default` the user has the same experience and output (one line of code will not make their life miserable) but we have many more cases covered due to these same fact, which means, less work in the future for us. More important, we align with different tastes and ideas.

For me that resonates more with us but I'm a democrat and I stick to what the majority decides! 😂

cc: @juanitorduz @wd60622

Comment (wd60622) [reply pymc_marketing/model_config.py:401]: Think having the centered=False key in the dictionary is an easy user experience, makes most sense to me, and is my preferred API. I am in favor of keeping the keys of kwargs close to actually distribution parameters because that is predictable in my eyes which makes me not like the "offset" option.

I've never heard of a non-centered hierarchical Beta distribution. Please dig one up if I you have a reference! @carlosagostini 

Personally think that the hierarchical support already is quite flexible! We do support 0-1 distributions and positive distributions with the current implementation so people do have options! For example: 
```python
# hierarchical beta example
positive_distribution = {"dist": "HalfNormal, "kwargs": {"alpha": 1, "beta": 1}
hierarchical_beta_distribution = {
    "dist": "Beta", 
    "kwargs": {"alpha": positive_distribution, "beta": positive_distribution},
    "dims": "channel", 
}
# hierarchical gamma example
hierarchical_positive_distribution = {
    "name": "Gamma", 
    "kwargs": {"alpha", positive_distribution, "beta": positive_distribution},
    "dims": "channel",
}
```
Another alternative to fix the domain could be that we add is a "transform" key. For instance, transform="exp" would be performed to after the variable is created. 
For example, 
```python
mu = {"dist": "Normal", "kwargs": {"mu": 0, "sigma": 1}}
non_centered_hierarchical_positive_distribution = {
    "name": "Normal", 
    "kwargs": {"mu", mu, "sigma": positive_distribution},
    "dims": "channel",
    "centered": False,
    "transform": "exp",
}
```
And the `transform` key would have to come from `pt` or `pm.math` or similar namespace. This might be inlined with non-centered but support different domains. [Ref](https://discourse.mc-stan.org/t/non-centered-parameterization-for-likelihood/31023/2)

Comment (juanitorduz) [reply pymc_marketing/model_config.py:401]: I think this is the key aspect: I also did not find any reference about non-centered parameterization for general distributions and also ended up in the exact Stan blog 😄

Note that Bob Carpenter (and he is really an expert) says:

> The beta distribution isn’t a location/scale family, so the notion of centering/non-centering doesn’t apply directly.

Hence, in view of this, I also like the centered bean and transform API suggested by @wd60622 .

My biggest concern with the offset option is that this could potentially mess up the geometry and hurt the sampler. Plus making the parameters unidentifiable because of too many degrees of freedom.

Comment (cetagostini) [reply pymc_marketing/model_config.py:401]: I was talking about "beta" like coefficient, "beta_channel" or a parameter in specific, definitely not sure about beta non center parametrization.

Nevertheless, I hear you loud and clear :) I'll make the change, and push soon.

**Comment by cetagostini** [comment]
Hey requested changes are done, and tests are working properly!
cc: @wd60622 @juanitorduz

File: pymc_marketing/model_config.py [hunk 3]
```diff
@@ -369,10 +383,66 @@ def handle_parameter_distributions(
...
+
+    sigma_global = create_distribution(
+        f"{name}_sigma_global",
+        sigma_dist["dist"],
+        sigma_dist["kwargs"],
+        dims=sigma_dims,
+    )
+
+    return pm.Deterministic(
+        name=name, var=mu_global + offset * sigma_global, dims=desired_dims
```
Comment (wd60622) [inline pymc_marketing/model_config.py:437]: Gotta use the dim_hander or broadcasting can break

File: pymc_marketing/model_config.py [hunk 4]
```diff
@@ -369,10 +383,66 @@ def handle_parameter_distributions(
...
+
+    """
+    desired_dims = kwargs.get("dims", ())
+
+    mu_dist = distribution_kwargs["mu"]
+    mu_dims = mu_dist.get("dims", ())
+    sigma_dist = distribution_kwargs["sigma"]
+    sigma_dims = sigma_dist.get("dims", ())
+
+    offset = pm.Normal(name="offset", mu=0, sigma=2, dims=desired_dims)
```
Comment (wd60622) [inline pymc_marketing/model_config.py:420]: sigma=1

Comment (wd60622) [inline pymc_marketing/model_config.py:420]: sigma=1

Comment (wd60622) [inline pymc_marketing/model_config.py:420]: name=f"{name}_offset"

File: pymc_marketing/model_config.py [hunk 5]
```diff
@@ -369,10 +383,66 @@ def handle_parameter_distributions(
...
+
+    mu_global = create_distribution(
+        f"{name}_mu_global",
```
Comment (wd60622) [inline pymc_marketing/model_config.py:423]: {name}_mu is how centered is so use that instead

File: pymc_marketing/model_config.py [hunk 6]
```diff
@@ -369,10 +383,66 @@ def handle_parameter_distributions(
...
+        mu_dist["dist"],
+        mu_dist["kwargs"],
+        dims=mu_dims,
+    )
+
+    sigma_global = create_distribution(
+        f"{name}_sigma_global",
```
Comment (wd60622) [inline pymc_marketing/model_config.py:430]: Same as above

File: pymc_marketing/model_config.py [hunk 7]
```diff
@@ -369,10 +383,66 @@ def handle_parameter_distributions(
...
+    desired_dims = kwargs.get("dims", ())
+
+    mu_dist = distribution_kwargs["mu"]
+    mu_dims = mu_dist.get("dims", ())
+    sigma_dist = distribution_kwargs["sigma"]
+    sigma_dims = sigma_dist.get("dims", ())
+
+    offset = pm.Normal(name="offset", mu=0, sigma=2, dims=desired_dims)
+
+    mu_global = create_distribution(
```
Comment (wd60622) [inline pymc_marketing/model_config.py:422]: What happens if mu includes a nested distribution? Doesnt seem to be handled

File: pymc_marketing/model_config.py [hunk 8]
```diff
@@ -392,9 +462,11 @@ def create_distribution(
     -------
     TensorVariable
         A PyMC random variable.
-
     """
-    dim_handler = create_dim_handler(kwargs.get("dims"))
+    if centered_hierarchy is False:
```
Comment (wd60622) [inline pymc_marketing/model_config.py:466]: if not centered_hierarchy:

File: pymc_marketing/model_config.py [hunk 9]
```diff
@@ -369,10 +383,66 @@ def handle_parameter_distributions(
...
+    return pm.Deterministic(
+        name=name, var=mu_global + offset * sigma_global, dims=desired_dims
+    )
+
+
 def create_distribution(
     name: str,
     distribution_name: str,
     distribution_kwargs: dict[str, Any],
+    centered_hierarchy: bool | None = None,
```
Comment (wd60622) [inline pymc_marketing/model_config.py:445]: I prefer name centered because it is shorter. 

Just make bool = True as default

File: tests/test_model_config.py [hunk 10]
```diff
@@ -224,6 +257,25 @@ def coords() -> dict[str, list[str]]:
...
+            ],
+            [(3, 2), (3,), (2,)],
+        ),
+        (
+            "hierarchical_non_centered",
+            [
+                "hierarchical_non_centered",
+                "hierarchical_non_centered_mu_global",
+                "hierarchical_non_centered_sigma_global",
+                "offset",
```
Comment (wd60622) [inline tests/test_model_config.py:275]: Add the prefix

File: tests/test_model_config.py [hunk 11]
```diff
@@ -224,6 +257,25 @@ def coords() -> dict[str, list[str]]:
...
+        (
+            "hierarchical_centered",
+            [
+                "hierarchical_centered",
+                "hierarchical_centered_mu",
+                "hierarchical_centered_sigma",
+            ],
+            [(3, 2), (3,), (2,)],
+        ),
+        (
```
Comment (wd60622) [inline tests/test_model_config.py:269]: Can you test a 2D version with centered=False

File: pymc_marketing/model_config.py (hunk 8 above)
Comment (wd60622) [inline pymc_marketing/model_config.py:466]: Need to check that dist=Normal and because that is assumed

**Review by wd60622 — CHANGES_REQUESTED** [review]

File: pymc_marketing/model_config.py (hunk 3 above)
Comment (cetagostini) [reply pymc_marketing/model_config.py:437]: Solve!

File: pymc_marketing/model_config.py (hunk 4 above)
Comment (cetagostini) [reply pymc_marketing/model_config.py:420]: solve!

File: pymc_marketing/model_config.py (hunk 5 above)
Comment (cetagostini) [reply pymc_marketing/model_config.py:423]: solve!

File: pymc_marketing/model_config.py (hunk 6 above)
Comment (cetagostini) [reply pymc_marketing/model_config.py:430]: solve!

File: pymc_marketing/model_config.py (hunk 9 above)
Comment (cetagostini) [reply pymc_marketing/model_config.py:445]: I manage cases by default is `True` if exist. We must have default `None` otherwise when loop is `if not None` the values pass and mess up everything.

```python
centered_flag = parameter_config.get("centered", True)
```

File: pymc_marketing/model_config.py (hunk 8 above)
Comment (cetagostini) [reply pymc_marketing/model_config.py:466]: That generates problem when is `None` and we need to handle the case when is none in case don't exist in the dict because they are not using hierarchies. Related to next comments!

File: tests/test_model_config.py (hunk 10 above)
Comment (cetagostini) [reply tests/test_model_config.py:275]: Done!

File: pymc_marketing/model_config.py (hunk 7 above)
Comment (cetagostini) [reply pymc_marketing/model_config.py:422]: Done!

File: pymc_marketing/model_config.py (hunk 4 above)
Comment (cetagostini) [reply pymc_marketing/model_config.py:420]: Done!

**Comment by wd60622** [comment]
Should we continue with this or go with #759

File: pymc_marketing/model_config.py (hunk 8 above)
Comment (cetagostini) [reply pymc_marketing/model_config.py:466]: Done!

**Comment by cetagostini** [comment]
I address all your comments @wd60622. My suggestion would be merge to have the feature, which probably will improve with #759 but users can start from now to play with it and give some feedback.

File: tests/test_model_config.py (hunk 11 above)
Comment (wd60622) [reply tests/test_model_config.py:269]: @cetagostini

File: tests/test_model_config.py [hunk 12]
```diff
@@ -224,6 +257,25 @@ def coords() -> dict[str, list[str]]:
...
         ("delta", ["delta"], [(3, 1)]),
+        (
+            "hierarchical_centered",
+            [
+                "hierarchical_centered",
+                "hierarchical_centered_mu",
+                "hierarchical_centered_sigma",
+            ],
+            [(3, 2), (3,), (2,)],
+        ),
```
Comment (wd60622) [inline tests/test_model_config.py:268]: Similar to this one but with non centered

**Review by wd60622 — CHANGES_REQUESTED** [review]
Just one more check then looks good

File: pymc_marketing/model_config.py [hunk 13]
```diff
@@ -392,9 +506,13 @@ def create_distribution(
     -------
     TensorVariable
         A PyMC random variable.
-
     """
-    dim_handler = create_dim_handler(kwargs.get("dims"))
+    if centered is False:
+        if distribution_name != "Normal":
+            NonCenterInvalidDistributionError(distribution_name)
```
Comment (wd60622) [inline pymc_marketing/model_config.py:512]: Need to raise this

**Review by wd60622 — CHANGES_REQUESTED** [review]

File: pymc_marketing/model_config.py (hunk 13 above)
Comment (cetagostini) [reply pymc_marketing/model_config.py]: My bad done!
